File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    visited_col = [False] * n
    visited_first_d = [False] * 2 * n # x+y가 같은 대각선
    visited_second_d = [False] * 2 * n # x-y+n-1가 같은 대각선
    
    answer = [0]
    
    def back_tracking(x):        
        if x == n:
            answer[0] += 1
            return
        
        for y in range(n):
            if visited_col[y] or visited_first_d[x+y] or visited_second_d[x-y+n-1]:
                continue
            
            visited_col[y] = True
            visited_first_d[x+y] = True
            visited_second_d[x-y+n-1] = True
            
            back_tracking(x+1)
            
            visited_col[y] = False
            visited_first_d[x+y] = False
            visited_second_d[x-y+n-1] = False
    
    back_tracking(0)
    
    return answer[0]

File: 프로그래머스/2/12952. N－Queen/N－Queen.js
function solution(n) {
    const visitedCol = Array(n).fill(false);
    const visitedFirstD = Array(2*n).fill(false); // x+y가 동일한 대각선
    const visitedSecondD = Array(2*n).fill(false); // x-y+n-1이 동일한 대각선
    
    let answer = 0;
    
    const backTracking = (x) => {
        if(x === n){
            answer++;
            return;
        }
        
        for(let y=0;y<n;y++){ // y위치의 col 계산
            if(visitedCol[y] || visitedFirstD[x+y] || visitedSecondD[x-y+n-1]) continue;
            // 해당하는 열과 대각선 방문 처리
            visitedCol[y] = true;
            visitedFirstD[x+y] = true;
            visitedSecondD[x-y+n-1] = true;
            
            backTracking(x+1); // 다음행으로 이동
            
            // 방문 처리 취소
            visitedCol[y] = false;
            visitedFirstD[x+y] = false;
            visitedSecondD[x-y+n-1] = false;
        }
    }
    
    backTracking(0)
    
    return answer
}